src/Compilers/CSharp/Portable/Binder/RefSafetyAnalysis.cs

```diff
@@ -0,0 +1,914 @@
+                    _diagnostics);
+            }
+
+            return null;
+        }
+
+        public override BoundNode? VisitFunctionPointerInvocation(BoundFunctionPointerInvocation node)
+        {
+            using var _ = GetArgumentPlaceholders(receiverOpt: null, node.Arguments);
+            base.VisitFunctionPointerInvocation(node);
+
+            if (!node.HasErrors)
+            {
+                var method = node.FunctionPointer.Signature;
+                CheckInvocationArgMixing(
+                    node.Syntax,
+                    method,
+                    receiverOpt: null,
+                    method.Parameters,
+                    node.Arguments,
+                    node.ArgumentRefKindsOpt,
+                    argsToParamsOpt: default,
+                    _localScopeDepth,
+                    _diagnostics);
+            }
+
+            return null;
+        }
+
+        public override BoundNode? VisitAwaitExpression(BoundAwaitExpression node)
+        {
+            var placeholders = ArrayBuilder<(BoundValuePlaceholderBase, uint)>.GetInstance();
+            GetAwaitableInstancePlaceholders(placeholders, node.AwaitableInfo, GetValEscape(node.Expression, _localScopeDepth));
+            using var _ = new PlaceholderRegion(this, placeholders);
+            placeholders.Free();
+            base.VisitAwaitExpression(node);
+            return null;
+        }
+
+        private void GetAwaitableInstancePlaceholders(ArrayBuilder<(BoundValuePlaceholderBase, uint)> placeholders, BoundAwaitableInfo awaitableInfo, uint valEscapeScope)
+        {
+            if (awaitableInfo.AwaitableInstancePlaceholder is { } placeholder)
+            {
+                placeholders.Add((placeholder, valEscapeScope));
+            }
+        }
+
+        // Based on NullableWalker.VisitDeconstructionAssignmentOperator().
+        public override BoundNode? VisitDeconstructionAssignmentOperator(BoundDeconstructionAssignmentOperator node)
+        {
+            base.VisitDeconstructionAssignmentOperator(node);
+
+            var left = node.Left;
+            var right = node.Right;
+            var variables = GetDeconstructionAssignmentVariables(left);
+            VisitDeconstructionArguments(variables, right.Syntax, right.Conversion, right.Operand);
+            variables.FreeAll(v => v.NestedVariables);
+            return null;
+        }
+
+        private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> variables, SyntaxNode syntax, Conversion conversion, BoundExpression right)
+        {
+            Debug.Assert(conversion.Kind == ConversionKind.Deconstruction);
+
+            // We only need to visit the right side when deconstruction uses a Deconstruct() method call
+            // (when !DeconstructionInfo.IsDefault), not when the right side is a tuple, because ref structs
+            // cannot be used as tuple type arguments.
+            if (!conversion.DeconstructionInfo.IsDefault)
```

Comment:
nit: consider moving this check into `VisitDeconstructionMethodArguments` as a fail-fast. Then `VisitDeconstructionArguments` could be removed